File: clients/java/openfeature-provider/src/main/java/io/juspay/superposition/provider/ConfigData.java
package io.juspay.superposition.provider;

import java.time.Instant;
import java.util.List;
import java.util.Map;

/**
 * Data class representing Superposition configuration data.
 */
public class ConfigData {
    private final Map<String, Object> defaultConfigs;
    private final List<Map<String, Object>> contexts;
    private final Map<String, Object> overrides;
    private final Map<String, Object> dimensions;
    private final Instant fetchedAt;

    public ConfigData(
        Map<String, Object> defaultConfigs,
        List<Map<String, Object>> contexts,
        Map<String, Object> overrides,
        Map<String, Object> dimensions
    ) {
        this.defaultConfigs = defaultConfigs;
        this.contexts = contexts;
        this.overrides = overrides;
        this.dimensions = dimensions;
        this.fetchedAt = Instant.now();
    }

    public Map<String, Object> getDefaultConfigs() {
        return defaultConfigs;
    }

    public List<Map<String, Object>> getContexts() {
        return contexts;
    }

    public Map<String, Object> getOverrides() {
        return overrides;
    }

    public Map<String, Object> getDimensions() {
        return dimensions;
    }

    public Instant getFetchedAt() {
        return fetchedAt;
    }
}

File: clients/java/openfeature-provider/src/main/kotlin/io/juspay/superposition/openfeature/data_source/ConfigData.kt
package io.juspay.superposition.openfeature.data_source

import uniffi.superposition_types.Config
import java.time.Instant

/**
 * Holds resolved configuration along with the timestamp it was fetched.
 *
 * Wraps [Config] (the UniFFI-generated Rust type) with a [fetchedAt] timestamp used to support
 * conditional fetching (HTTP 304 / If-Modified-Since). For HTTP sources this is the server's
 * `last-modified` time, not the local clock, so it can be sent straight back as
 * `if-modified-since`.
 *
 * [Config] contains:
 * - `contexts` — list of contexts matchable against an evaluation context
 * - `overrides` — per-context key/value overrides (`Map<String, Overrides>`)
 * - `defaultConfigs` — default key/value pairs (`ExtendedMap = Map<String, String>`)
 * - `dimensions` — dimension definitions (`Map<String, DimensionInfo>`)
 *
 * @property data The resolved configuration, directly from the UniFFI Rust bindings.
 * @property fetchedAt The timestamp when this config was last modified at its source.
 */
data class ConfigData(val data: Config, val fetchedAt: Instant) {

    override fun toString(): String =
        "ConfigData(fetched_at: $fetchedAt" +
            ", data.contexts: ${data.contexts.size}" +
            ", data.overrides: ${data.overrides.size}" +
            ", data.defaultConfigs: ${data.defaultConfigs.size}" +
            ", data.dimensions: ${data.dimensions.size})"
}
